massa-final-state/src/final_state.rs

```diff
@@ -25,6 +27,10 @@ use tracing::{debug, info, warn};
 use massa_models::config::PERIODS_BETWEEN_BACKUPS;
 use massa_models::timeslots::get_block_slot_timestamp;
 
+// TODO when creating a final ledger at genesis, initialize the execution hash trail to Hash::zero()
+
+// TODO after bootstrap, check that the execution hash trail is present and in correct format
```

Comment:
This would be done in the is_db_valid() method at the end of this file.
If the db iterator encounters your added (serialized_key, serialized_value), it should not return false.
(And you can also add a separate check that the key exists in the DB, although I never check this currently).

Reply:
I think that things could be simplified a bit and better encapsulated:
* remove the is_db_valid function completely
* after the bootstrap of the db (or its initialization or loading from disk) is done, create final_state by passing it the db and letting it initialize all its subsystems in its constructor: 
  * pos_state = PosState::try_from_db(&db).map_err(||e| format!("could not init PosState in FinalState: {}"), e)?;
  * async_pool = AsyncPool::try_from_db(&db).map_err(||e| format!("could not AsyncPool state in FinalState: {}"), e)?;
  * ... each of those constructors initializes the subsystem but also checks the db validity on the prefix subset(s) they manage

That way we have each module take responsibility for checking its own parts of the ledger properly and we reduce the error surface when adding new elements to the db (eg. forgetting to add them to is_db_valid)

Reply:
I agree things could be improved!
I didn't want to change too much things at one time, but I agree avoiding initialization of the final state before the bootstrap would be a better workflow.

Here is the issue: https://github.com/massalabs/massa/issues/4234#issue-1794866153

Reply:
Thanks https://github.com/massalabs/massa/issues/4234